pkg/test/framework/runtime/dependency/creation_processor.go

```diff
@@ -17,153 +17,230 @@ package dependency
 import (
 	"fmt"
 	"reflect"
+	"time"
 
 	"istio.io/istio/pkg/test/framework/api/component"
 	"istio.io/istio/pkg/test/framework/api/lifecycle"
-	"istio.io/istio/pkg/test/framework/runtime/key"
 )
 
 // creationProcessor is used by Manager to resolve creation order for components.
 type creationProcessor struct {
-	scope    lifecycle.Scope
-	mgr      *Manager
-	provided map[component.ID]component.Descriptor
-	required map[key.Instance]*requiredEntry
+	scope lifecycle.Scope
+	mgr   *Manager
+
+	// The entries that need to be processed.
+	required map[namedId]*reqEntry
 }
 
 // newCreationProcessor creates a new creation processor for the Manager.
 func newCreationProcessor(mgr *Manager, scope lifecycle.Scope) *creationProcessor {
 	return &creationProcessor{
 		scope:    scope,
 		mgr:      mgr,
-		provided: make(map[component.ID]component.Descriptor),
-		required: make(map[key.Instance]*requiredEntry),
+		required: make(map[namedId]*reqEntry),
 	}
 }
 
-type requiredEntry struct {
-	desc component.Descriptor
-	ids  map[component.ID]bool
+// A struct representing a named ID. This is used as the key to what requirements need to be
+// created, as the Variant is not important, just the ID and Name.
+type namedId struct {
```

Comment:
I see here that you're explicitly using component ID. Previously, a "requirement" could be either a Descriptor or a component ID, so that the user could say that I want "Pilot" (component ID) or more specifically the "RealPilot" (descriptor). Is that still the case? And if so, do you think that still makes sense?

With this work, I feel like we may need to rethink our terminology a bit to make things more clear. Here's what we have:

- **component.ID** - specifies the ID of the `interface` that the component exposes. Every component instance can only expose 1 interface.
- **component.Descriptor** - this describes a particular variant that exposes the component interface (e.g. fake vs real). Currently we don't really use variant, but the thought was that we might in integration tests. Variant is similar to an implementation of an interface, except that you have the extra dimension of environment (k8s vs native), so variant + environment = implementation.
- **namedId** - an identifier for an instance of a component (but does not specify variant/descriptor)

I wonder if the following changes might help:

- **component.ID** -> **component.InterfaceID**
- **component.Descriptor** -> **component.Type**
- **namedId** -> **component.ID**

WDYT?

Reply:
So that's definitely still the case, and this type is actually just an internal type and not exposed to the user. It is only used to track where we have named IDs that the user has requested, so we can load the defaults for those.

I actually think the current naming is fine, although I don't love the name ConfiguredRequirement that I came up with, need something better there :)

But yeah I think it is actually OK as-is for the most part.